src/hooks/useTable.tsx

```diff
@@ -0,0 +1,196 @@
+import {
+  type TypedDocumentNode as DocumentNode,
+  useQuery,
+} from '@apollo/client';
+import type { DataGridProps } from '@mui/x-data-grid';
+import {
+  type FieldNode,
+  getOperationAST,
+  Kind,
+  type SelectionSetNode,
+} from 'graphql';
+import { get, pick, set, uniqBy } from 'lodash';
+import { useMemo, useState } from 'react';
+import type { Get, Paths, SetNonNullable } from 'type-fest';
+import {
+  isNetworkRequestInFlight,
+  type PaginatedListInput,
+  type PaginatedListOutput,
+  type SortableListInput,
+} from '~/api';
+import type { Order } from '~/api/schema/schema.graphql';
+import { lowerCase, upperCase } from '~/common';
+
+type ListInput = SetNonNullable<
+  Required<SortableListInput & PaginatedListInput>
+>;
+
+type PathsMatching<T, List> = {
+  [K in Paths<T>]: K extends string
+    ? Get<T, K> extends List
+      ? K
+      : never
+    : never;
+}[Paths<T>];
+
+const defaultInitialInput = {
+  page: 1,
+  count: 25,
+  sort: 'id',
+  order: 'ASC' as Order,
+};
+const defaultKeyArgs = ['__typename', 'id'];
+
+export const useTable = <
+  Output extends Record<string, any>,
+  Vars extends { input?: ListInput; [s: string]: any },
+  const Path extends PathsMatching<Output, PaginatedListOutput<any>> & string,
+  List extends PaginatedListOutput<any> = Get<Output, Path> extends infer U
+    ? U extends PaginatedListOutput<any>
+      ? U
+      : never
+    : never
+>({
+  query,
+  variables,
+  listAt,
+  initialInput,
+  keyArgs = defaultKeyArgs,
+}: {
+  query: DocumentNode<Output, Vars>;
+  variables: Vars;
+  listAt: Path;
+  initialInput?: Partial<ListInput>;
+  keyArgs?: string[];
+}) => {
+  const [input, onChange] = useState(() => ({
+    ...defaultInitialInput,
+    ...initialInput,
+  }));
+
+  const queryForItemRef = useMemo(() => {
+    const queryForItemRef = structuredClone(query);
```

Comment:
Do we want to shadow like this?

Reply:
I don't have a problem with shadowing per se. Both of these variables represent the same thing, it's just the outer one is the result of the memoization. This is `useMemo` not `useCallback`, so the type of both is the same thing, never a function.

I still think this code is clear and straight forward. That doesn't mean other code that involves shadowing could be confusing, like if it shadows something completely different.